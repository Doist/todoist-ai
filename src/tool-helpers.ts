import {
    type MoveTaskArgs,
    type PersonalProject,
    type Task,
    type TodoistApi,
    type WorkspaceProject,
    getSanitizedContent,
} from '@doist/todoist-api-typescript'
import z from 'zod'
import { formatDuration } from './utils/duration-parser.js'

export type Project = PersonalProject | WorkspaceProject

export function isPersonalProject(project: Project): project is PersonalProject {
    return 'inboxProject' in project
}

export function isWorkspaceProject(project: Project): project is WorkspaceProject {
    return 'accessLevel' in project
}

/**
 * Creates a MoveTaskArgs object from move parameters, validating that exactly one is provided.
 * @param taskId - The task ID (used for error messages)
 * @param projectId - Optional project ID to move to
 * @param sectionId - Optional section ID to move to
 * @param parentId - Optional parent ID to move to
 * @returns MoveTaskArgs object with exactly one destination
 * @throws Error if multiple move parameters are provided or none are provided
 */
export function createMoveTaskArgs(
    taskId: string,
    projectId?: string,
    sectionId?: string,
    parentId?: string,
): MoveTaskArgs {
    // Validate that only one move parameter is provided (RequireExactlyOne constraint)
    const moveParams = [projectId, sectionId, parentId].filter(Boolean)
    if (moveParams.length > 1) {
        throw new Error(
            `Task ${taskId}: Only one of projectId, sectionId, or parentId can be specified at a time. The Todoist API requires exactly one destination for move operations.`,
        )
    }

    if (moveParams.length === 0) {
        throw new Error(
            `Task ${taskId}: At least one of projectId, sectionId, or parentId must be provided for move operations.`,
        )
    }

    // Build moveArgs with the single defined value
    if (projectId) return { projectId }
    if (sectionId) return { sectionId }
    if (parentId) return { parentId }

    // This should never be reached due to the validation above
    throw new Error('Unexpected error: No valid move parameter found')
}

/**
 * Map a single Todoist task to a more structured format, for LLM consumption.
 * @param task - The task to map.
 * @returns The mapped task.
 */
function mapTask(task: Task) {
    return {
        id: task.id,
        content: getSanitizedContent(task.content),
        description: getSanitizedContent(task.description),
        dueDate: task.due?.date,
        recurring: task.due?.isRecurring && task.due.string ? task.due.string : false,
        priority: task.priority,
        projectId: task.projectId,
        sectionId: task.sectionId,
        parentId: task.parentId,
        labels: task.labels,
        duration: task.duration ? formatDuration(task.duration.amount) : null,
    }
}

/**
 * Map a single Todoist project to a more structured format, for LLM consumption.
 * @param project - The project to map.
 * @returns The mapped project.
 */
function mapProject(project: Project) {
    return {
        id: project.id,
        name: project.name,
        color: project.color,
        isFavorite: project.isFavorite,
        isShared: project.isShared,
        parentId: isPersonalProject(project) ? (project.parentId ?? null) : null,
        inboxProject: isPersonalProject(project) ? (project.inboxProject ?? false) : false,
        viewStyle: project.viewStyle,
    }
}

const ErrorSchema = z.object({
    httpStatusCode: z.number(),
    responseData: z.object({
        error: z.string(),
        errorCode: z.number(),
        errorTag: z.string(),
    }),
})

async function getTasksByFilter({
    client,
    query,
    limit,
    cursor,
}: {
    client: TodoistApi
    query: string
    limit: number | undefined
    cursor: string | undefined
}) {
    try {
        const { results, nextCursor } = await client.getTasksByFilter({
            query,
            cursor,
            limit,
        })

        return {
            tasks: results.map(mapTask),
            nextCursor,
        }
    } catch (error) {
        const parsedError = ErrorSchema.safeParse(error)
        if (!parsedError.success) {
            throw error
        }
        const { responseData } = parsedError.data
        if (responseData.errorTag === 'INVALID_SEARCH_QUERY') {
            throw new Error(`Invalid filter query: ${query}`)
        }
        throw new Error(
            `${responseData.error} (tag: ${responseData.errorTag}, code: ${responseData.errorCode})`,
        )
    }
}

export { getTasksByFilter, mapTask, mapProject }
