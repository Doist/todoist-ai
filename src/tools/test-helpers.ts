import type { PersonalProject, Section, Task } from '@doist/todoist-api-typescript'

/**
 * Mapped task type matching the output of mapTask function.
 * Used for filter-based query test mocks.
 */
export type MappedTask = {
    id: string
    content: string
    description: string
    dueDate: string | undefined
    recurring: string | boolean
    priority: number
    projectId: string
    sectionId: string | null
    parentId: string | null
    labels: string[]
}

/**
 * Creates a mock Task with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockTask(overrides: Partial<Task> = {}): Task {
    return {
        id: '8485093748',
        content: 'Test task content',
        description: '',
        completedAt: null,
        labels: [],
        childOrder: 1,
        priority: 1,
        projectId: '6cfCcrrCFg2xP94Q',
        sectionId: null,
        parentId: null,
        url: 'https://todoist.com/showTask?id=8485093748',
        // Use correct property names from Task schema
        noteCount: 0,
        addedByUid: '713437',
        addedAt: '2025-08-13T22:09:56.123456Z',
        deadline: null,
        responsibleUid: null,
        assignedByUid: null,
        isCollapsed: false,
        isDeleted: false,
        duration: null,
        checked: false,
        updatedAt: '2025-08-13T22:09:56.123456Z',
        due: null,
        dayOrder: 0,
        userId: '713437',
        ...overrides,
    }
}

/**
 * Creates a mock Section with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockSection(overrides: Partial<Section> = {}): Section {
    return {
        id: 'section-123',
        projectId: '6cfCcrrCFg2xP94Q',
        sectionOrder: 1,
        userId: 'test-user',
        addedAt: '2024-01-01T00:00:00Z',
        updatedAt: '2024-01-01T00:00:00Z',
        archivedAt: null,
        isArchived: false,
        isDeleted: false,
        isCollapsed: false,
        name: 'Test Section',
        ...overrides,
    }
}

/**
 * Creates a mock PersonalProject with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockProject(overrides: Partial<PersonalProject> = {}): PersonalProject {
    return {
        id: '6cfCcrrCFg2xP94Q',
        name: 'Test Project',
        color: 'charcoal',
        isFavorite: false,
        isShared: false,
        parentId: null,
        inboxProject: false,
        viewStyle: 'list',
        url: 'https://todoist.com/projects/6cfCcrrCFg2xP94Q',
        isDeleted: false,
        updatedAt: '2025-08-13T22:09:55.841800Z',
        createdAt: '2025-08-13T22:09:55.841785Z',
        childOrder: 1,
        defaultOrder: 0,
        description: '',
        isCollapsed: false,
        canAssignTasks: false,
        isFrozen: false,
        isArchived: false,
        ...overrides,
    }
}

/**
 * Creates a mock API response object with results and nextCursor.
 */
export function createMockApiResponse<T>(
    results: T[],
    nextCursor: string | null = null,
): {
    results: T[]
    nextCursor: string | null
} {
    return {
        results,
        nextCursor,
    }
}

/**
 * Creates a simplified mapped task (matches mapTask output) for filter-based query tests.
 */
export function createMappedTask(overrides: Partial<MappedTask> = {}): MappedTask {
    return {
        id: TEST_IDS.TASK_1,
        content: 'Test task content',
        description: '',
        dueDate: undefined,
        recurring: false,
        priority: 1,
        projectId: TEST_IDS.PROJECT_TEST,
        sectionId: null,
        parentId: null,
        labels: [],
        ...overrides,
    }
}

/**
 * Common error messages used across tests.
 */
export const TEST_ERRORS = {
    API_RATE_LIMIT: 'API Error: Rate limit exceeded',
    API_UNAUTHORIZED: 'API Error: Unauthorized',
    INVALID_CURSOR: 'Invalid cursor format',
    INVALID_FILTER: 'Invalid filter query',
} as const

/**
 * Creates multiple test cases for parameterized testing.
 */
export function createTestCases<T, E = unknown>(
    cases: Array<{ name: string; input: T; expected?: E }>,
) {
    return cases
}

/**
 * Common mock IDs used across tests for consistency.
 */
export const TEST_IDS = {
    TASK_1: '8485093748',
    TASK_2: '8485093749',
    TASK_3: '8485093750',
    PROJECT_INBOX: 'inbox-project-id',
    PROJECT_WORK: 'work-project-id',
    PROJECT_TEST: '6cfCcrrCFg2xP94Q',
    SECTION_1: 'section-123',
    SECTION_2: 'section-456',
    USER_ID: '713437',
} as const
