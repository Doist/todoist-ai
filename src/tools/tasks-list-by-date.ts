import { addDays, formatISO } from 'date-fns'
import { z } from 'zod'
import { getToolOutput } from '../mcp-helpers.js'
import type { TodoistTool } from '../todoist-tool.js'
import { getTasksByFilter } from '../tool-helpers.js'
import { generateTaskNextSteps, previewTasks, summarizeList } from '../utils/response-builders.js'
import { TOOL_NAMES } from '../utils/tool-names.js'

const ArgsSchema = {
    startDate: z
        .string()
        .regex(/^(\d{4}-\d{2}-\d{2}|today|overdue)$/)
        .describe(
            "The start date to get the tasks for. Format: YYYY-MM-DD, 'today', or 'overdue'.",
        ),
    daysCount: z
        .number()
        .int()
        .min(1)
        .max(30)
        .default(1)
        .describe(
            "The number of days to get the tasks for, starting from the start date. Ignored when startDate is 'overdue'.",
        ),
    limit: z
        .number()
        .int()
        .min(1)
        .max(50)
        .default(10)
        .describe('The maximum number of tasks to return.'),
    cursor: z
        .string()
        .optional()
        .describe(
            'The cursor to get the next page of tasks (cursor is obtained from the previous call to this tool, with the same parameters).',
        ),
}

const tasksListByDate = {
    name: TOOL_NAMES.TASKS_LIST_BY_DATE,
    description:
        "Get tasks by date range or overdue tasks. Use startDate 'overdue' for overdue tasks, or provide a date/date range.",
    parameters: ArgsSchema,
    async execute(args, client) {
        let query: string

        if (args.startDate === 'overdue') {
            query = 'overdue'
        } else {
            const startDate =
                args.startDate === 'today'
                    ? formatISO(new Date(), { representation: 'date' })
                    : args.startDate
            const endDate = addDays(startDate, args.daysCount + 1)
            const endDateStr = formatISO(endDate, { representation: 'date' })
            query = `(due after: ${startDate} | due: ${startDate}) & due before: ${endDateStr}`
        }

        const result = await getTasksByFilter({
            client,
            query,
            cursor: args.cursor,
            limit: args.limit,
        })

        const textContent = generateTextContent({
            tasks: result.tasks,
            args,
            nextCursor: result.nextCursor,
        })

        return getToolOutput({
            textContent,
            structuredContent: {
                tasks: result.tasks,
                nextCursor: result.nextCursor,
                totalCount: result.tasks.length,
                hasMore: Boolean(result.nextCursor),
                appliedFilters: {
                    startDate: args.startDate,
                    daysCount: args.daysCount,
                    limit: args.limit,
                    cursor: args.cursor,
                },
            },
        })
    },
} satisfies TodoistTool<typeof ArgsSchema>

function generateTextContent({
    tasks,
    args,
    nextCursor,
}: {
    tasks: Awaited<ReturnType<typeof getTasksByFilter>>['tasks']
    args: z.infer<z.ZodObject<typeof ArgsSchema>>
    nextCursor: string | null
}) {
    // Generate filter description
    const filterHints: string[] = []
    if (args.startDate === 'overdue') {
        filterHints.push('overdue tasks only')
    } else if (args.startDate === 'today') {
        filterHints.push(`today${args.daysCount > 1 ? ` + ${args.daysCount - 1} more days` : ''}`)
    } else {
        filterHints.push(
            `${args.startDate}${args.daysCount > 1 ? ` to ${addDays(args.startDate, args.daysCount).toISOString().split('T')[0]}` : ''}`,
        )
    }

    // Generate subject description
    const subject =
        args.startDate === 'overdue'
            ? 'Overdue tasks'
            : args.startDate === 'today'
              ? `Today's tasks`
              : `Tasks for ${args.startDate}`

    // Generate helpful suggestions for empty results
    const zeroReasonHints: string[] = []
    if (tasks.length === 0) {
        if (args.startDate === 'overdue') {
            zeroReasonHints.push('Great job! No overdue tasks')
            zeroReasonHints.push("Check today's tasks with startDate='today'")
        } else {
            zeroReasonHints.push("Expand date range with larger 'daysCount'")
            zeroReasonHints.push("Check 'overdue' for past-due items")
        }
    }

    // Generate contextual next steps
    const now = new Date()
    const todayStr = now.toISOString().split('T')[0]
    const nextSteps = generateTaskNextSteps('listed', tasks, {
        hasToday: args.startDate === 'today' || tasks.some((task) => task.dueDate === todayStr),
        hasOverdue:
            args.startDate === 'overdue' ||
            tasks.some((task) => task.dueDate && new Date(task.dueDate) < now),
    })

    return summarizeList({
        subject,
        count: tasks.length,
        limit: args.limit,
        nextCursor: nextCursor ?? undefined,
        filterHints,
        previewLines: previewTasks(tasks),
        zeroReasonHints,
        nextSteps,
    })
}

export { tasksListByDate }
