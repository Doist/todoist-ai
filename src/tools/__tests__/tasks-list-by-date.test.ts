import type { TodoistApi } from '@doist/todoist-api-typescript'
import { jest } from '@jest/globals'
import { getTasksByFilter } from '../../tool-helpers.js'
import { tasksListByDate } from '../tasks-list-by-date.js'
import { TEST_ERRORS, TEST_IDS, createMappedTask, extractTextContent } from '../test-helpers.js'

// Mock the tool helpers
jest.mock('../../tool-helpers', () => ({
    getTasksByFilter: jest.fn(),
}))

const mockGetTasksByFilter = getTasksByFilter as jest.MockedFunction<typeof getTasksByFilter>

// Mock the Todoist API (not directly used by tasks-list-by-date, but needed for type)
const mockTodoistApi = {} as TodoistApi

// Mock date-fns functions to make tests deterministic
jest.mock('date-fns', () => ({
    addDays: jest.fn(() => new Date('2025-08-16')), // Return predictable end date
    formatISO: jest.fn((date, options) => {
        if (typeof date === 'string') {
            return date // Return string dates as-is
        }
        if (
            options &&
            typeof options === 'object' &&
            'representation' in options &&
            options.representation === 'date'
        ) {
            return '2025-08-15' // Return predictable date for 'today'
        }
        return '2025-08-16' // Return predictable end date
    }),
}))

describe('tasks-list-by-date tool', () => {
    beforeEach(() => {
        jest.clearAllMocks()
        // Mock current date to make tests deterministic
        jest.spyOn(Date, 'now').mockReturnValue(new Date('2025-08-15T10:00:00Z').getTime())
    })

    afterEach(() => {
        jest.restoreAllMocks()
    })

    describe('listing overdue tasks', () => {
        it.each([
            { daysCount: 7, hasTasks: true, description: 'with tasks' },
            { daysCount: 5, hasTasks: false, description: 'ignoring daysCount' },
        ])('should handle overdue tasks $description', async ({ daysCount, hasTasks }) => {
            const mockTasks = hasTasks
                ? [
                      createMappedTask({
                          id: TEST_IDS.TASK_1,
                          content: 'Overdue task',
                          dueDate: '2025-08-10',
                          priority: 2,
                          labels: ['urgent'],
                      }),
                  ]
                : []

            const mockResponse = { tasks: mockTasks, nextCursor: null }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            const result = await tasksListByDate.execute(
                { startDate: 'overdue', limit: 50, daysCount },
                mockTodoistApi,
            )

            expect(mockGetTasksByFilter).toHaveBeenCalledWith({
                client: mockTodoistApi,
                query: 'overdue',
                cursor: undefined,
                limit: 50,
            })
            // Verify result is a concise summary
            expect(extractTextContent(result)).toMatchSnapshot()

            // Verify structured content
            const { structuredContent } = result
            expect(structuredContent.tasks).toHaveLength(hasTasks ? 1 : 0)
            expect(structuredContent.totalCount).toBe(hasTasks ? 1 : 0)
            expect(structuredContent.hasMore).toBe(false)
            expect(structuredContent.nextCursor).toBeNull()
            expect(structuredContent.appliedFilters.startDate).toBe('overdue')
        })
    })

    describe('listing tasks by date range', () => {
        it('should get tasks for today when startDate is "today"', async () => {
            const mockTasks = [createMappedTask({ content: 'Today task', dueDate: '2025-08-15' })]
            const mockResponse = { tasks: mockTasks, nextCursor: null }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            const result = await tasksListByDate.execute(
                { startDate: 'today', limit: 50, daysCount: 7 },
                mockTodoistApi,
            )

            expect(mockGetTasksByFilter).toHaveBeenCalledWith({
                client: mockTodoistApi,
                query:
                    expect.stringContaining('due after:') && expect.stringContaining('due before:'),
                cursor: undefined,
                limit: 50,
            })
            // Verify result is a concise summary
            expect(extractTextContent(result)).toMatchSnapshot()
        })

        it.each([
            {
                name: 'specific date',
                params: { startDate: '2025-08-20', limit: 50, daysCount: 7 },
                tasks: [createMappedTask({ content: 'Specific date task', dueDate: '2025-08-20' })],
                cursor: null,
            },
            {
                name: 'multiple days with pagination',
                params: {
                    startDate: '2025-08-20',
                    daysCount: 3,
                    limit: 20,
                    cursor: 'current-cursor',
                },
                tasks: [
                    createMappedTask({
                        id: TEST_IDS.TASK_2,
                        content: 'Multi-day task 1',
                        dueDate: '2025-08-20',
                    }),
                    createMappedTask({
                        id: TEST_IDS.TASK_3,
                        content: 'Multi-day task 2',
                        dueDate: '2025-08-21',
                    }),
                ],
                cursor: 'next-page-cursor',
            },
        ])('should handle $name', async ({ params, tasks, cursor }) => {
            const mockResponse = { tasks, nextCursor: cursor }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            const result = await tasksListByDate.execute(params, mockTodoistApi)

            expect(mockGetTasksByFilter).toHaveBeenCalledWith({
                client: mockTodoistApi,
                query: expect.stringContaining('2025-08-20'),
                cursor: params.cursor || undefined,
                limit: params.limit,
            })
            // Verify result is a concise summary
            expect(extractTextContent(result)).toMatchSnapshot()
        })
    })

    describe('pagination and limits', () => {
        it.each([
            {
                name: 'pagination parameters',
                params: {
                    startDate: 'today',
                    limit: 25,
                    daysCount: 7,
                    cursor: 'pagination-cursor',
                },
                expectedCursor: 'pagination-cursor',
                expectedLimit: 25,
            },
            {
                name: 'default values',
                params: { startDate: '2025-08-15', limit: 50, daysCount: 7 },
                expectedCursor: undefined,
                expectedLimit: 50,
            },
        ])('should handle $name', async ({ params, expectedCursor, expectedLimit }) => {
            const mockResponse = { tasks: [], nextCursor: null }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            await tasksListByDate.execute(params, mockTodoistApi)

            expect(mockGetTasksByFilter).toHaveBeenCalledWith({
                client: mockTodoistApi,
                query: expect.any(String),
                cursor: expectedCursor,
                limit: expectedLimit,
            })
        })
    })

    describe('edge cases', () => {
        it.each([
            { name: 'empty results', daysCount: 7, shouldReturnResult: true },
            { name: 'maximum daysCount', daysCount: 30, shouldReturnResult: false },
            { name: 'minimum daysCount', daysCount: 1, shouldReturnResult: false },
        ])('should handle $name', async ({ daysCount, shouldReturnResult }) => {
            const mockResponse = { tasks: [], nextCursor: null }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            const startDate = daysCount === 7 ? 'today' : '2025-08-15'
            const result = await tasksListByDate.execute(
                { startDate, limit: 50, daysCount },
                mockTodoistApi,
            )

            expect(mockGetTasksByFilter).toHaveBeenCalledTimes(1)
            if (shouldReturnResult) {
                // Verify result is a concise summary
                expect(extractTextContent(result)).toMatchSnapshot()
            }
        })
    })

    describe('next steps logic', () => {
        it('should suggest appropriate actions when hasOverdue is true', async () => {
            const mockTasks = [
                createMappedTask({
                    id: TEST_IDS.TASK_1,
                    content: 'Overdue task from list',
                    dueDate: '2025-08-10', // Past date - creates hasOverdue context
                }),
            ]
            const mockResponse = { tasks: mockTasks, nextCursor: null }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            const result = await tasksListByDate.execute(
                { startDate: '2025-08-15', limit: 10, daysCount: 1 },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain(
                'Use tasks-update-multiple to modify priorities or due dates',
            )
        })

        it('should suggest today-focused actions when startDate is today', async () => {
            const mockTasks = [
                createMappedTask({
                    id: TEST_IDS.TASK_1,
                    content: "Today's task",
                    dueDate: '2025-08-15', // Today's date based on our mock
                }),
            ]
            const mockResponse = { tasks: mockTasks, nextCursor: null }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            const result = await tasksListByDate.execute(
                { startDate: 'today', limit: 10, daysCount: 1 },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain(
                'Use tasks-update-multiple to modify priorities or due dates',
            )
        })

        it('should suggest appropriate actions when startDate is overdue', async () => {
            const mockTasks = [
                createMappedTask({
                    id: TEST_IDS.TASK_1,
                    content: 'Overdue task',
                    dueDate: '2025-08-10',
                }),
            ]
            const mockResponse = { tasks: mockTasks, nextCursor: null }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            const result = await tasksListByDate.execute(
                { startDate: 'overdue', limit: 10, daysCount: 1 },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain(
                'Use tasks-update-multiple to modify priorities or due dates',
            )
        })

        it('should provide helpful suggestions for empty overdue results', async () => {
            const mockResponse = { tasks: [], nextCursor: null }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            const result = await tasksListByDate.execute(
                { startDate: 'overdue', limit: 10, daysCount: 1 },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Great job! No overdue tasks')
            expect(textContent).toContain("Check today's tasks with startDate='today'")
        })

        it('should provide helpful suggestions for empty date range results', async () => {
            const mockResponse = { tasks: [], nextCursor: null }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            const result = await tasksListByDate.execute(
                { startDate: '2025-08-20', limit: 10, daysCount: 1 },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain("Expand date range with larger 'daysCount'")
            expect(textContent).toContain("Check 'overdue' for past-due items")
        })
    })

    describe('error handling', () => {
        it.each([
            {
                error: TEST_ERRORS.INVALID_FILTER,
                params: { startDate: 'today', limit: 50, daysCount: 7 },
            },
            {
                error: TEST_ERRORS.API_RATE_LIMIT,
                params: { startDate: 'overdue', limit: 50, daysCount: 7 },
            },
            {
                error: TEST_ERRORS.INVALID_CURSOR,
                params: {
                    startDate: '2025-08-15',
                    limit: 50,
                    daysCount: 7,
                    cursor: 'invalid-cursor',
                },
            },
        ])('should propagate $error', async ({ error, params }) => {
            mockGetTasksByFilter.mockRejectedValue(new Error(error))
            await expect(tasksListByDate.execute(params, mockTodoistApi)).rejects.toThrow(error)
        })
    })
})
