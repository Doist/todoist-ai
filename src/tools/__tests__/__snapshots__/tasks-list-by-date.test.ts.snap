// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`tasks-list-by-date tool edge cases should handle empty results 1`] = `
"Today's tasks: 0 (limit 50).
Filter: today + 6 more days.
No results. Expand date range with larger 'daysCount'; Check 'overdue' for past-due items."
`;

exports[`tasks-list-by-date tool listing overdue tasks should handle overdue tasks ignoring daysCount 1`] = `
"Overdue tasks: 0 (limit 50).
Filter: overdue tasks only.
No results. Great job! No overdue tasks; Check today's tasks with startDate='today'."
`;

exports[`tasks-list-by-date tool listing overdue tasks should handle overdue tasks with tasks 1`] = `
"Overdue tasks: 1 (limit 50).
Filter: overdue tasks only.
Preview:
    Overdue task • due 2025-08-10 • P2 • id=8485093748
Next:
- Use tasks-update-multiple to modify priorities or due dates
- Use tasks-complete-multiple to mark finished tasks
- Focus on overdue items first to get back on track"
`;

exports[`tasks-list-by-date tool listing tasks by date range should get tasks for today when startDate is "today" 1`] = `
"Today's tasks: 1 (limit 50).
Filter: today + 6 more days.
Preview:
    Today task • due 2025-08-15 • P1 • id=8485093748
Next:
- Use tasks-update-multiple to modify priorities or due dates
- Use tasks-complete-multiple to mark finished tasks
- Focus on overdue items first to get back on track"
`;

exports[`tasks-list-by-date tool listing tasks by date range should handle multiple days with pagination 1`] = `
"Tasks for 2025-08-20: 2 (limit 20), more available.
Filter: 2025-08-20 to 2025-08-16.
Preview:
    Multi-day task 1 • due 2025-08-20 • P1 • id=8485093749
    Multi-day task 2 • due 2025-08-21 • P1 • id=8485093750
Next:
- Use tasks-update-multiple to modify priorities or due dates
- Use tasks-complete-multiple to mark finished tasks
- Pass cursor 'next-page-cursor' to fetch more results."
`;

exports[`tasks-list-by-date tool listing tasks by date range should handle specific date 1`] = `
"Tasks for 2025-08-20: 1 (limit 50).
Filter: 2025-08-20 to 2025-08-16.
Preview:
    Specific date task • due 2025-08-20 • P1 • id=8485093748
Next:
- Use tasks-update-multiple to modify priorities or due dates
- Use tasks-complete-multiple to mark finished tasks"
`;

exports[`tasks-list-by-date tool next steps logic should provide helpful suggestions for empty date range results 1`] = `
"Tasks for 2025-08-20: 0 (limit 10).
Filter: 2025-08-20.
No results. Expand date range with larger 'daysCount'; Check 'overdue' for past-due items."
`;

exports[`tasks-list-by-date tool next steps logic should provide helpful suggestions for empty overdue results 1`] = `
"Overdue tasks: 0 (limit 10).
Filter: overdue tasks only.
No results. Great job! No overdue tasks; Check today's tasks with startDate='today'."
`;

exports[`tasks-list-by-date tool next steps logic should suggest appropriate actions when hasOverdue is true 1`] = `
"Tasks for 2025-08-15: 1 (limit 10).
Filter: 2025-08-15.
Preview:
    Overdue task from list • due 2025-08-10 • P1 • id=8485093748
Next:
- Use tasks-update-multiple to modify priorities or due dates
- Use tasks-complete-multiple to mark finished tasks
- Focus on overdue items first to get back on track"
`;

exports[`tasks-list-by-date tool next steps logic should suggest appropriate actions when startDate is overdue 1`] = `
"Overdue tasks: 1 (limit 10).
Filter: overdue tasks only.
Preview:
    Overdue task • due 2025-08-10 • P1 • id=8485093748
Next:
- Use tasks-update-multiple to modify priorities or due dates
- Use tasks-complete-multiple to mark finished tasks
- Focus on overdue items first to get back on track"
`;

exports[`tasks-list-by-date tool next steps logic should suggest today-focused actions when startDate is today 1`] = `
"Today's tasks: 1 (limit 10).
Filter: today.
Preview:
    Today's task • due 2025-08-15 • P1 • id=8485093748
Next:
- Use tasks-update-multiple to modify priorities or due dates
- Use tasks-complete-multiple to mark finished tasks
- Focus on overdue items first to get back on track"
`;
